### ISSUE TITLE
[proposal] using prototypes & forward declarations for target architecture descriptions

### ISSUE THREAD

**Issue by smolkaj** [p4lang/p4-spec#74]
I think that standard C-style [prototypes](https://en.wikipedia.org/wiki/Function_prototype) and [forward declarations](https://en.wikipedia.org/wiki/Forward_declaration) could be used to great effect in P4 for target architecture descriptions. Let me demonstrate by example:

``` C
/* v1model.p4 */
...
// Architecture
parser Parser<H, M>(packet_in b,
                    out H parsedHdr,
                    inout M meta,
                    inout standard_metadata_t standard_metadata);
control VerifyChecksum<H, M>(in H hdr,
                             inout M meta,
                             inout standard_metadata_t standard_metadata);
control Ingress<H, M>(inout H hdr,
                      inout M meta,
                      inout standard_metadata_t standard_metadata);
control Egress<H, M>(inout H hdr,
                     inout M meta,
                     inout standard_metadata_t standard_metadata);
control ComputeChecksum<H, M>(inout H hdr,
                              inout M meta,
                              inout standard_metadata_t standard_metadata);
control Deparser<H>(packet_out b, in H hdr);

package V1Switch<H, M>(Parser<H, M> p,
                       VerifyChecksum<H, M> vr,
                       Ingress<H, M> ig,
                       Egress<H, M> eg,
                       ComputeChecksum<H, M> ck,
                       Deparser<H> dep
                       );
```

Above you see part of `v1mode.p4`. It specifies which components must be implemented by the programmer to obtain a complete configuration for this target. One may call this the "hardware programming interface" of the target. I propose the following approach to declare the same interface, using [forward declarations](https://en.wikipedia.org/wiki/Forward_declaration):

``` C
// Architecture
struct H; // the user must define a struct containing the headers
struct M; // the user must define a struct containing the meta fields

parser Parser(packet_in b,
              out H parsedHdr,
              inout M meta,
              inout standard_metadata_t standard_metadata);
control VerifyChecksum(in H hdr,
                       inout M meta,
                       inout standard_metadata_t standard_metadata);
control Ingress(inout H hdr,
                inout M meta,
                inout standard_metadata_t standard_metadata);
control Egress(inout H hdr,
               inout M meta,
               inout standard_metadata_t standard_metadata);
control ComputeChecksum(inout H hdr,
                        inout M meta,
                        inout standard_metadata_t standard_metadata);
control Deparser(packet_out b, in H hdr);

// no need to declare a package
```

Let me point out the main changes:
- The architecture description [declares](http://stackoverflow.com/questions/1410563/what-is-the-difference-between-a-definition-and-a-declaration) structs `H` and `M`, but [does not define](http://stackoverflow.com/questions/1410563/what-is-the-difference-between-a-definition-and-a-declaration) them. The user will have to provide these definitions.
- The `parser` and `control` declarations are not interpreted as type declarations as before, but rather as [prototype](https://en.wikipedia.org/wiki/Function_prototype) declarations. The user will have to provide implementations for these [prototypes](https://en.wikipedia.org/wiki/Function_prototype).

This leads to some simplifications:
- The `parser` and `control` declarations are no longer polymorphic. (They have monomorphic types involving `H` and `M`.)
- The concept of a `package` is no longer needed. Its sole purpose (as far as I know) was to enforce certain type constraints, which now happens "for free".

We also gain some extra expressivity:
- The architecture actually requires `H` and `M` to be structs (`p4c-bm2-ss` throws an error if you use, for example, a primitive type). Now we can actually express that in the language. Thus, the user can immediately see that constraint and it can be enforced through type checking by the front end. This is much cleaner than throwing a (potentially cryptic) error message in the backend.

With this proposal, architecture descriptions would be very much like header files in C, or abstract classes in Java. This is what a user program would look like:

``` C
#include<v1model.p4>

struct { ... } H;
struct { ... } M;

parser Parser(packet_in b, out H parsedHdr, inout M meta, inout standard_metadata_t std_meta) {
  ...
}
...
// no need to declare main
```

Note that unlike before, the user can (and must!) define a `parser` with name `Parser`. Currently this would lead to an error, because `Parser` is interpreted as a type that was already defined in `v1model.p4`.

To summarize, my proposal
- uses a well-known concept from C (forward declarations), rather than an ad-hoc concept introduced by P4
- eliminates the need for package declarations
- eliminates the need for polymorphism in certain places
- raises the expressivity of the language, allowing
  - the user to see certain constraints in the type signature
  - the compiler to catch certain illegal programs through type checking in the front end
- requires only modest modifications to the current language

I am curious to hear how others feel about this proposal, and what potential downsides may be.
